Split Rough Spec

AksharP5 opened the pull request:
```
refactor: split and validate rough_spec before parsing

Split the rough_spec into a RoughSpec object that holds attributes for the three distinct sections: lexical, syntactic, and semantic. Add a validation method, which returns a list of ValidationErrors. This will make it easier to parse each individual part later on.

---

* Closes #4
* Closes #12

---

Co-authored-by: Akshar Patel <aksharpatel1233@gmail.com>
Co-authored-by: Stoney Jackson <dr.stoney@gmail.com>

```

AksharP5 commented:
@StoneyJackson 

The split_rough_spec function has been created, along with the respective unit tests. I have also added a test for the load_rough_spec, to ensure that everything is functional at the end when a file is passed. This PR should be ready for review, if there is anything that looks wrong or is missing, please let me know. Also, if you need me to update the commit message, please inform me.

# How To Test
* Open this PR in GitPod
* Run 'bin/test/units.bash'
* Ensure all tests pass

The next question I have is regarding when the RoughSpec object that is returned will be sent to the parsing sections, should that also be done in one of the files for the load_rough_spec section or will that get handled elsewhere? Also, will each section only be receiving their respective attribute, or the entire object?

AksharP5 commented:
@StoneyJackson do you want me to accommodate for #12 in this PR and add all block objects to the different sections even if it shouldn't be there, so that it can be validated afterwards? Would you also like for me to get rid of the TypeError that is being thrown all together, as it should not be an issue as long as the other parts of the code are working as intended? Also, for the ValidationError, if the checks are only for blocks, should the line being appended be the first Line object in the Block?

AksharP5 commented:
@StoneyJackson The validation of the rough_spec has been implemented, and necessary tests have been created, everything should be all set, the only thing I can think of that may be unnecessary is the TypeError that I am throwing and its related test, however it does not hurt to keep it, in my opinion.

StoneyJackson reviewed (changes requested):
OK, I think I made some design errors. Let's separate the validator from the rough spec. The validator should know about the rough spec, but not the other way around.

Please do the following:

1. [x] Move the validator and its tests to separate modules (`validate_rough_spec.py` and `validate_rough_spec_test.py`).
2. [x] Remove the errorList attribute from RoughSpec, and move the error class to the validator.
3. [x] Modify validate_rough_spec and its checks to return a list of errors rather than appending them to the RoughSpec's errorList (which it no longer has).

While we are at it, please also

1. [x] Rename attributes of RoughSpec to use camel casing instead of snake casing. This was the other mistake I mentioned during our meeting.
2. [x] Try to refactor split_rough_spec into smaller functions/methods. This will likely require creating a class (maybe `RoughSpecSplitter`) to house the new methods so they can share state, and not have to pass around a bunch of parameters. Then `split_rough_spec` instantiates this class, and calls its entrypoint method (probably `split`).

AksharP5 commented:
@StoneyJackson the changes you have requested have been made, please let me know if there is anything missing, or if there are any other issues

StoneyJackson commented on src/plcc/load_spec/load_rough_spec/split_rough_spec.py:
```diff
@@ -0,0 +1,83 @@
+    semanticSectionList: list[list[Line | Divider | Block]]
+
+#Class to handle the splitting of the rough_spec
+class RoughSpecSplitter:
```
Let's see if we can simplify this more.

First, the only element type that interacts with this algorithm is Divider. Anything other than a Divider is passed through to the current section. This observation should allow you to avoid the TypeError and will prevent your algorithm from breaking if a new element type is added.

Next, let's try to avoid boolean flags. Boolean flags create an indirection between the condition you want to test and make the code harder to follow. In this case, you're using them to track the current section to append to. What if you had a variable that always stored the current section, and swap this out for a new list when a Divider is found. But then you'll still need to figure out when section (lex, syn, sems) to store it in. So how about use a list of sections.

    sectionList = []

To start a new section, append a list to it.

    sectionList.append([])

Now the current section is always at the end of the list. To append an element to the current list append it to the last section.

    sectionList[-1].append(element)

When the sorting is done, you should be able to extract the appropriate sections from this list to build RoughSpec.

Give this another go and see if this helps beat down the complexity and length of this implementation.

AksharP5 replied:
The changes have been made, please let me know if there is anything else

StoneyJackson reviewed (changes requested):
The design is much better. Thank you!

I think we're almost there. See my in-code comment.

You're doing great! Keep up the good work!

StoneyJackson commented:
@AksharP5 Please read the commit messages of the last three commits I made. Then please clean up validate_rough_spec_test mimicking what I did in split_rough_spec_test.

Feel free to squash all of these without giving me credit. I just found it easier to give you an example rather than explaining it all.

AksharP5 commented:
@StoneyJackson the file has been cleaned up, also I added strip() to some of the helper functions, as I double checked the modified tests, and there were line objects being made from empty strings, specifically in test_no_divider(), but everything should be working now, please let me know if there is anything else to refactor.

StoneyJackson commented:
@AksharP5 I removed more decencies from the test. But it required significant rewrites. Sorry.

Squash and I think we're good.

AksharP5 commented on src/plcc/load_spec/load_rough_spec/load_rough_spec_test.py:
Does this test file also need to be refactored, and if so how should it be done?

AksharP5 commented on src/plcc/load_spec/load_rough_spec/validate_rough_spec_test.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def test_non_blocks_do_not_cause_an_error():
+    block = makeEmptyBlock()
```
Should I remove this from this test, as it is not being used?

AksharP5 committed:
refactor: split and validate rough_spec before parsing

Split the rough_spec into a RoughSpec object that holds attributes for the three distinct sections: lexical, syntactic, and semantic. Add a validation method, which returns a list of ValidationErrors. This will make it easier to parse each individual part later on.

---

* Closes #4
* Closes #12

---

Co-authored-by: Akshar Patel <aksharpatel1233@gmail.com>
Co-authored-by: Stoney Jackson <dr.stoney@gmail.com>

AksharP5 commented:
Has been squashed and co-author credits have been provided.

> @AksharP5 I removed more decencies from the test. But it required significant rewrites. Sorry.
> 
> Squash and I think we're good.

StoneyJackson commented:
Sold! Thank you!
